### ISSUE TITLE
Add NTS support (ntpsec) [Request]

### ISSUE THREAD

**Issue by siga1975** [opnsense/plugins#2162]
It would be nice to have NTS in place as an alternative to common NTP

reference: []https://blog.apnic.net/2019/11/08/network-time-security-new-ntp-authentication-mechanism/

Here a PoC I did on 2 raspberry pi 4

```
root@time1:/root # grep -Ev '^#|^$' /etc/ntp.conf
driftfile /var/lib/ntpsec/ntp.drift
leapfile /usr/share/zoneinfo/leap-seconds.list
statsdir /var/log/ntpsec/
nts enable
nts key /etc/ntpsec/nts-cert/time1.signorini.in.key
nts cert /etc/ntpsec/nts-cert/time1.signorini.in.crt
nts cookie /var/lib/ntp/nts-keys
refclock shm unit 0 refid SHM minpoll 4 maxpoll 4 time1 +0.040 noselect
refclock shm unit 1 prefer refid GPGA minpoll 4 maxpoll 4
server time0.signorini.in iburst minpoll 4 maxpoll 4 nts
server time2.signorini.in iburst minpoll 4 maxpoll 4
server time.cloudflare.com:1234 iburst minpoll 6 nts
server time1.ethz.ch minpoll 7 iburst
server time2.ethz.ch minpoll 7 iburst
```

```
root@time1:/root # /usr/local/bin/ntpq -p
     remote                                   refid      st t when poll reach   delay   offset   jitter
=======================================================================================================
 SHM(0)                                  .SHM.            0 l   10   16  377   0.0000  -4.5807   3.4257
*SHM(1)                                  .GPGA.           0 l    9   16  377   0.0000   0.0289   0.0007
+time0.signorini.in                      .GPS.            1 8    2   16  377   1.0362  -0.0211   0.0134
+time2.signorini.in                      .GNSS.           1 u    1   16  377   0.3531   0.0561   0.0172
 time.cloudflare.com                     .NTS.           16 5    -  128    0   0.0000   0.0000   0.0019
-time1.ethz.ch                           .PPS.            1 u   60  128    3  12.7593   0.4802   0.2725
-time2.ethz.ch                           .PPS.            1 u   62  128    3  13.4803   0.4194   0.1739
```

```
root@time1:/root # cat /var/log/syslog|grep ntpd|grep -3 TLS|tail -15
May 21 19:08:45  time1  ntpd[1334]: NTSc: DNS lookup of time0.signorini.in took 0.002 sec
May 21 19:08:45  time1  ntpd[1334]: NTSc: nts_probe connecting to time0.signorini.in:123 => 192.168.33.41:123
May 21 19:08:45  time1  ntpd[1334]: NTSc: set cert host: time0.signorini.in
May 21 19:08:46  time1  ntpd[1334]: NTSc: Using TLSv1.3, TLS_AES_256_GCM_SHA384 (256)
May 21 19:08:46  time1  ntpd[1334]: NTSc: certificate subject name: /CN=time0.signorini.in
May 21 19:08:46  time1  ntpd[1334]: NTSc: certificate issuer name: /C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
May 21 19:08:46  time1  ntpd[1334]: NTSc: certificate is valid.
--
May 21 19:08:47  time1  ntpd[1334]: NTSc: DNS lookup of time.cloudflare.com:1234 took 0.037 sec
May 21 19:08:47  time1  ntpd[1334]: NTSc: nts_probe connecting to time.cloudflare.com:1234 => 162.159.200.1:123
May 21 19:08:47  time1  ntpd[1334]: NTSc: set cert host: time.cloudflare.com
May 21 19:08:47  time1  ntpd[1334]: NTSc: Using TLSv1.3, TLS_AES_256_GCM_SHA384 (256)
May 21 19:08:47  time1   ntpd[1334]: NTSc: certificate subject name: /C=US/ST=California/L=San Francisco/O=Cloudflare, Inc./CN=time.cloudflare.com
May 21 19:08:47  time1  ntpd[1334]: NTSc: certificate issuer name: /C=US/O=DigiCert Inc/CN=DigiCert ECC Secure Server CA
May 21 19:08:47  time1  ntpd[1334]: NTSc: certificate is valid.
```

**Comment by psychogun** [comment]
Hi. I might have spoken too soon; atleast, in some regards. Questions in **bold**.

I am coming from PFSense, just because of this feature (NTS, NTP over TLS). So today I installed OPNsense, and the Chrony plugin. 

On the configuration page of Chrony, you are unable to add a custom port - reading this blog post here; if you have a NTS enabled client, you should point it to `time.cloudflare.com:1234` - https://blog.cloudflare.com/secure-time/ 

> If you have an NTS client, point it at time.cloudflare.com:1234. Otherwise point your NTP client at time.cloudflare.com. More details on configuration are available in the developer docs.

And if you try this in the GUI, `time.cloudflare.com:1234`, you get:
> Please correct validation errors in form 

Allthough, that blog post might just be a bit old. Because if I check `NTS Client` on the configuration page and just use `time.cloudflare.com` - it works:

```
$ chronyc sources
MS Name/IP address         Stratum Poll Reach LastRx Last sample               
===============================================================================
^* time.cloudflare.com           3   6   177    63   +161us[+1006us] +/- 4242us

$ sudo chronyc -N authdata
Password:
Name/IP address             Mode KeyID Type KLen Last Atmp  NAK Cook CLen
=========================================================================
time.cloudflare.com          NTS     1   15  256  349    0    0    8  100
```
Under `Mode` it says `NTS`, so the quest is actually fullfilled, time has been securely synchronized. 

Without NTS checked:
```
$ sudo chronyc -N authdata
Password:
Name/IP address             Mode KeyID Type KLen Last Atmp  NAK Cook CLen
=========================================================================
time.cloudflare.com            -     0    0    0    -    0    0    0    0
```

So. Should we / or should we not be able **to define a Peer with a custo**m port?
Not that important, especially now that `time.cloudflare.com` seems to work, in this case. 

```
$ chronyc tracking
Reference ID    : A29FC87B (time.cloudflare.com)
Stratum         : 4
Ref time (UTC)  : Tue Dec 29 22:45:48 2020
System time     : 0.010503863 seconds fast of NTP time
Last offset     : -0.000681546 seconds
RMS offset      : 0.000681546 seconds
Frequency       : 112.957 ppm slow
Residual freq   : -419.687 ppm
Skew            : 1000000.000 ppm
Root delay      : 0.008276920 seconds
Root dispersion : 13.921768188 seconds
Update interval : 2.0 seconds
Leap status     : Normal
$ 
```

Anyhoo, it would be wonderful to have some sort of st**atus of the output from `chronyc` displayed in the GUI, as shown above. Am I synchronized? Am I synchronized, securely?**


And to continue; how can we actually benefit from this?
- Let us assume that the LAN(s) are secured- any NTP time (without NTS) found on the internet (on default port:123) is / has been compromised
- I am not willing to compile and install NTS enabled NTP clients on all my computers and devices; on many devices it will not even be possible.
- I want to continue to use the default port of 123 on all my LAN devices
- On specific devices / networks, I'll use DNS host overrides which would give me NTP from a LAN interface (let's assume time coming from the LAN is correct, no bother with NTS here)
- On other computers, I'll specify a NTP adress manually, which should/ could be an interface from the OPNsense firewall.

What should I do? Disable the "old" Network Time Server from Services? 
Then define all the networks I would like Chrony to serve / listen to, manually, in the little input box and change `Port` from `323` to `123`? Wha**t about a drop down menu here for the interfaces, such as in the Netwo**rk Time Server. 

<br>
I tried this as a test, but it did not work (Network Time Server would not synchronize to Chrony - it is getting late too, might have overseen a few things): 

Interfaces > Other Types > + Add
```
Parent interface: LAN
VLAN tag: 200
VLAN priority: Best Effort (0, default)
Description: Provide NTP from OPNsense
```
Save.

Interfaces > Assignments.
```
New interface: vlan 200 on re0 (Provide NTP from OPNsense)
Description: 200_NTP
```

Click + to add, then hit Save. 

Go `Interfaces` and select `[200_NTP]`. Enable the interface and set a static IPv4 adress. 

To make Chrony act as a server, add the static IPv4 adress you entered above, in the field marked `Allowed Networks` in the configuration page of Chrony. 

Now, since you are unable to define a custom port for a Time Server from the default Network Time in the GUI, under Services > Network Time > General - change the port in Chrony from `323` to `123`. Hit save. 

Go to GUI > Services > Network Time > General and add the static IPv4 adress above as a preferred `Time server`.

You should perhaps also define a random LAN interface under `Interface(s)` in this menu, to avoid port conflict (UDP 123 ??)

Hit Save. 

<br>
This did not work. Just stays on "INIT" in the `Status` page.
<br><br>
I feel there are a few bits and pieces which could be worked on with this Chrony plugin. .. Especially if you do not need PPS and stuffs. I found this over at PFSense feature request: https://redmine.pfsense.org/issues/8149

What are you guys thinking of regards to NTPsec? https://www.freshports.org/net/ntpsec


Other public NTS supported servers I have found over time and have written down, are (I do not guarantee if any of them still work);
nts.sth1.ntp.se
zoo.weinigel.se:4446
ntpmon.dcs1.biz
ntp2.glypnod.com
spidey.rellim.com
khronos.mikieboy.net
nts-test.strangled.net:443
nts3-e.ostfalia.de:443

Some useful links:
https://fedoraproject.org/wiki/Changes/NetworkTimeSecurity 
https://opensource.com/article/18/12/manage-ntp-chrony
https://www.thegeekdiary.com/centos-rhel-7-configuring-ntp-using-chrony/

**Comment by psychogun** [comment]
For now I have found out that it is OK to stop the Network Time Daemon, change port in Chrony to `123` and manually define the Allowed Networks.

It works as suggested; tested with `sntp -d address-to-firewall` on local LAN. 

What does not work, is _disabling_ Network Time Service. I can stop it, but whenever a reboot occour, it is running again.

**Comment by mimugmail** [comment]
@psychogun If you don't have any time servers configured for ntpd it shouldn't start at bootup.
Chrony uses by default port 323 to not disturb ntpd and still be able to run (which is default by chrony itself).
Currently I have no plans to add a port to the list for NTS, better use the ones which offer the service via standard port.
Regarding internal security you can redirect every connection from LAN to port 123 to local chrony, and plugin itself syncrhonize via NTS :)

I can add couple of diagnostics when I find the time, should be easy.

**Comment by psychogun** [comment]
@mimugmail, I removed the time servers - and ntpd is not starting at boot anymore. Thank you for that information, that was not intuitive :) (Perhaps should be included in the Help section?)

Please check my comment over at https://github.com/opnsense/core/issues/2012 - I have somewhat resolved my issue.

For Diagnostics, I have mentioned a few above - and how about including`chronyc clients`?

> This command shows a list of clients that have accessed the server,
           through the NTP, command, or NTS-KE port. It does not include
           accesses over the Unix domain command socket.

          

An example of the output is:
```


               Hostname                      NTP   Drop Int IntL Last     Cmd   Drop Int  Last
               ===============================================================================
               localhost                       2      0   2   -   133      15      0  -1     7
               foo.example.net                12      0   6   -    23       0      0   -     -
```

**Comment by psychogun** [comment]
Hi @mimugmail 
Just checking in, in case some other people reads this.

I was a bit confused about what you meant by
> Currently I have no plans to add a port to the list for NTS, better use the ones which offer the service via standard port.

As it was unclear to me what was / is the standard port for NTS, regards to the links I had provided earlier. I've found out, just recently actually,  as of Oct 2020 port number 4460 is the official port for NTS. 



https://www.netnod.se/time-and-frequency/netnod-launch-one-of-the-first-nts-enabled-time-services-in-the-world
> Update Oct 6, 2020: The NTS draft has been accepted as RFC8915.  The official port number for NTS is now 4460.

https://blog.cloudflare.com/nts-is-now-rfc/
> We’ve changed our time service to use the officially assigned port of 4460 for NTS key exchange, so you can use our service with ease.

https://www.rfc-editor.org/info/rfc8915 
> The NTS key establishment protocol is conducted via TCP port 4460.


And this is the default port used by Chrony if `nts` is enabled :) - https://chrony.tuxfamily.org/doc/devel/chrony.conf.html

> ntsport port
This option specifies the TCP port on which the server is listening for NTS-KE connections when the nts option is enabled. The default is 4460.


Thanks.
